src/mlpack/methods/ann/layer/weight_norm_impl.hpp

```diff
@@ -0,0 +1,168 @@
+      wrappedLayer);
+
+  output = boost::apply_visitor(outputParameterVisitor, wrappedLayer);
+}
+
+template<typename InputDataType, typename OutputDataType,
+         typename... CustomLayers>
+template<typename eT>
+void WeightNorm<InputDataType, OutputDataType, CustomLayers...>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, wrappedLayer)), std::move(gy), std::move(
+      boost::apply_visitor(deltaVisitor, wrappedLayer))), wrappedLayer);
+
+  g = boost::apply_visitor(deltaVisitor, wrappedLayer);
+}
+
+template<typename InputDataType, typename OutputDataType,
+         typename... CustomLayers>
+template<typename eT>
+void WeightNorm<InputDataType, OutputDataType, CustomLayers...>::Gradient(
+    arma::Mat<eT>&& input,
+    arma::Mat<eT>&& error,
+    arma::Mat<eT>&& gradient)
+{
+  ResetGradients(layerGradients);
+
+  // Calculate the gradients of the wrapped layer.
+  boost::apply_visitor(GradientVisitor(std::move(input),
+      std::move(error)), wrappedLayer);
+
+  // Store the norm of vector parameter temporarily.
+  double normVectorParameter = arma::norm(vectorParameter, 2);
+
+  // Set the gradients of the bias terms.
+  if (biasWeightSize != 0)
+  {
+    gradient.rows(0, biasWeightSize - 1) = arma::mat(layerGradients.memptr() +
+        layerWeightSize - biasWeightSize, biasWeightSize, 1, false, false);
+  }
+
+  // Calculate the gradients of the scalar parameter.
+  gradient[gradient.n_rows - 1] = arma::accu(layerGradients.rows(0,
+      layerWeightSize - biasWeightSize - 1) % vectorParameter) /
+      normVectorParameter;
+
+  // Calculate the gradients of the vector parameter.
+  gradient.rows(biasWeightSize, layerWeightSize - 1) =
+      scalarParameter(0) / normVectorParameter * (layerGradients.rows(0,
+      layerWeightSize - biasWeightSize - 1) - gradient[gradient.n_rows - 1] /
+      normVectorParameter * vectorParameter);
+}
+
+template<typename InputDataType, typename OutputDataType,
+         typename... CustomLayers>
+void WeightNorm<InputDataType, OutputDataType, CustomLayers...>::ResetGradients(
+    arma::mat& gradient)
+{
+    boost::apply_visitor(GradientSetVisitor(std::move(gradient), 0),
+        wrappedLayer);
+}
+
+template<typename InputDataType, typename OutputDataType,
+         typename... CustomLayers>
+template<typename Archive>
+void WeightNorm<InputDataType, OutputDataType, CustomLayers...>::serialize(
+    Archive& ar, const unsigned int /* version */)
+{
+  if (Archive::is_loading::value)
+  {
+    boost::apply_visitor(deleteVisitor, wrappedLayer);
+  }
+
+  ar & BOOST_SERIALIZATION_NVP(wrappedLayer);
+  ar & BOOST_SERIALIZATION_NVP(layerWeightSize);
+
+  // If we are loading, we need to initialize the weights.
+  if (Archive::is_loading::value)
+  {
+    weights.set_size(2 * layerWeightSize + 1, 1);
```

Comment:
Here you set a different size as compared to here:
https://github.com/mlpack/mlpack/pull/1920/files#diff-8627a19cad1a77c6ce8149cf66884073R34

Reply:
Thanks done.